### PR TITLE
First attempt at implementing #2404

### PR DESCRIPTION
**Pull request by pzinn**
This is a draft for now. Though it seems to work on my comp, I want to test how it compiles first before it can be merged.
- #2404 is implemented: most real functions now accept constants as arguments.
- I improved the expansion of `exp` into power series, so that this doesn't hang:
```
R=QQ[x]; exp x
```
and added a bunch more Taylor expansions of common functions.

### CODE REVIEW

**Comment by pzinn** [comment]
I forget how or if I can flag this as draft

**Comment by DanGrayson** [comment]
> I forget how or if I can flag this as draft

Me, too!

**Comment by mahrud** [comment]
I have a suggestion that may involve fewer changes:
- when exporting the function in `d`, make it unprotected:
https://github.com/Macaulay2/M2/blob/0b0159476fbb752b5c0bfd9a50a50b41c0cf95a5/M2/Macaulay2/d/regex.dd#L152
- then in top-level overwrite it:
https://github.com/Macaulay2/M2/blob/0b0159476fbb752b5c0bfd9a50a50b41c0cf95a5/M2/Macaulay2/m2/regex.m2#L33-L34

So you can revert all the changed in `d` and only edit these lines:
```diff
- setupfun("sin",sin);
+ setupfun("sin",sin).Protected = false;
```
and redefine them in `reals.m2`.

ps: where did `exp` go?!

**Comment by pzinn** [comment]
`exp` hasn't really changed, it was already `exp'`. your solution seems viable too, not sure which one is cleaner. ultimately the real issue is more about transferring all the methods from the old/primed version to the new/unprimed version of the function.

**Comment by pzinn** [comment]
Incidentally, the code for `exp RingElement` has an `error "expected an algebra over QQ"` but the example involves a polynomial ring over `ZZ/11`...

**Comment by pzinn** [comment]
I like better the idea of rededining the functions rather than having primed versions. I will give it a try.

**Comment by pzinn** [comment]
I'm not sure it will involve less work, though. The problem is, I mentioned above the methods still need to be transferred. If the functions are redefined, that means one needs to remove all the "typical values" from the d code. but then, how to guess which methods need to be bedeployed?

**Comment by pzinn** [comment]
I guess one option is to leave the typical values in d, and redefine the functions *after* typicalvalues.m2 is run...

File: M2/Macaulay2/d/actors3.d [hunk 1]
```diff
@@ -714,45 +714,45 @@ setup(StarS,starfun,timesfun);
 
 sin(e:Expr):Expr := (
      when e
-     is x:CCcell do toExpr(sin(x.v))				    -- # typical value: sin, CC, CC
-     is x:RRcell do toExpr(sin(x.v))				    -- # typical value: sin, RR, RR
-     is x:RRicell do toExpr(sin(x.v))				    -- # typical value: sin, RRi, RRi
-     is x:ZZcell do toExpr(sin(toRR(x.v)))			    -- # typical value: sin, ZZ, RR
-     is x:QQcell do toExpr(sin(toRR(x.v)))			    -- # typical value: sin, QQ, RR
+     is x:CCcell do toExpr(sin(x.v))				    -- # typical value: sin', CC, CC
```
[the remark this answers is not in the document]
Comment (pzinn) [reply M2/Macaulay2/d/actors3.d:717]: well, no. the methods of the actual `sin` won't be dummy functions, so they shouldn't be flagged as typical values.
that was (one of the reasons) why @mahrud suggested to not use primed functions, hence my second attempt.

[the remark this answers is not in the document]
Comment (mahrud) [reply M2/Macaulay2/d/actors3.d:717]: You can also dump the typical values in this implementation. `sin'` won't be exported, so who cares about it's typical values.

**Comment by pzinn** [comment]
I think I'll close this and focus on my 2nd attempt. I think there are too many potential issues with this 1st attempt.

**Pull request closed without merging**